torchaudio/prototype/tacotron2.py

```diff
@@ -0,0 +1,847 @@
+# *****************************************************************************
+#  Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#      * Redistributions of source code must retain the above copyright
+#        notice, this list of conditions and the following disclaimer.
+#      * Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in the
+#        documentation and/or other materials provided with the distribution.
+#      * Neither the name of the NVIDIA CORPORATION nor the
+#        names of its contributors may be used to endorse or promote products
+#        derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+#  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+#  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+#  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+#  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+#  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+#  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# *****************************************************************************
+
+from math import sqrt
+from typing import Tuple, List, Optional, Union
+
+import torch
+from torch import nn
+from torch import Tensor
+from torch.nn import functional as F
+
+
+__all__ = [
+    "Tacotron2",
+]
+
+
+class _LinearNorm(torch.nn.Module):
```

Comment:
I understand that this is being ported from Nvidia's code but I wonder if we can change this to a function. After all, what this class does is instantiate a class from PyTorch and initialize it with Xavier formula. This can be simply replaced with

```python
def _get_linear_norm(...):
    linear = torch.nn.Linear(...)
    torch.nn.init.xavier_uniform_(linear.weight, ...)
    return linear
```

Of course, the resulting model structure is slightly different and if you have a pre-trained model already, then a model surgery will be necessary to use it.
But in my opinion, it is more maintainable if there is less custom layers.

Reply:
I've changed it [this way](https://github.com/pytorch/audio/pull/1621/files#diff-8b8ec57192b9aaeb99a9ec5e709c977771b31b49fe7ddc6704c2128cf161a00eR42). Please let me know how you think :).

Reply:
Similarly, please do the same for `_ConvNorm` and replace by `_get_conv_norm` with a similar structure.

Reply:
Added [here](https://github.com/pytorch/audio/pull/1621/files#diff-8b8ec57192b9aaeb99a9ec5e709c977771b31b49fe7ddc6704c2128cf161a00eR64)